### ISSUE TITLE
disable sensor

### ISSUE THREAD

**Issue by mmoskal** [microsoft/jacdac#553]
Certain sensors consume significant power, for example various gas sensors. Should we add `enabled` register on case-by-case basis? Or should it maybe be an optional (default on) register on all sensors?

I'm also thinking about the color sensor - they sometimes have a backlight which consumes significant energy. An `enabled` bit may be useful as well.

**Comment by pelikhan** [comment]
sounds like a generic optional enable rw register is useful

**Comment by jamesadevine** [comment]
Might be a bit of a confusing dev ex? Either the client library would write enable, or it would be on the user to write enable before using? Maybe it should be intrinsic to the server that a write/read sets enabled?

Whichever way you cut it someone’s job gets a bit harder, and the implementation gets a bit muddier and the experience inconsistent.

James
________________________________

**Comment by mmoskal** [comment]
It also makes certain use cases possible - eg battery-powered button-operated color sensor (when you want to measure color of something and save it somewhere). 

I think power consumption is one of the features that distinguishes a prototyping system from a toy.

**Comment by jamesadevine** [comment]
I agree power consumption is important and I am in favour of an overall enable/disable model __but__ I don't think it should be optional. We should also specify the behaviour and implementation of enable/disable so that it is a consistent experience across hardware.

**Comment by mmoskal** [comment]
Another way to do it would be to add an optional "sleep" command and have any reading re-enable the sensor. Many of our sensors right now only enable on first reading.

One concern could be that jacdac-ts always (?) enables streaming, so if I wanted to write a vm program that, upon button press, takes one reading and sleeps the sensor, then I think jacdac-ts would wake it immediately because of streaming?

**Comment by pelikhan** [comment]
If we set the streaming interval properly, jacdac-ts will wait until asking for another reading.
________________________________

**Comment by mmoskal** [comment]
Current plan of record: 
* add optional enabled register, to be used mostly as `set enabled to false` aka `sleep`
* for devices that start disabled, they will (re-)enable themselves on first `get reading` of `set streaming_samples (to non 0)` - this is already current behavior for G0 potentiometer, rotary encoder, and env sensors
* devices that do not bother with enable/disable (most Padauks) just ignore the `enabled` register and are always enabled
* to deal with sensors that take time to initialize (my gas sensors has 15s startup, but they sometimes need half hour), we can use the `status_code` generic register (already specified, but never implemented)

This is consistent from the user perspective in the sense if they want to save power, they need to issue `sleep` command, which may be ignored. Otherwise they don't have to do anything special.

To better deal with with these optional registers (in general) we can consider #564

**Comment by tballmsft** [comment]
@mmoskal - can we close this?
